### ISSUE TITLE
demos/golang/run_golang_on_occlum.sh failed to run with 0.15.0-ubuntu18.04 container image

### ISSUE THREAD

**Issue by james-dong-security** [occlum/occlum#123]
After following all instructions documented at demos/golang/README.md, "run_golang_on_occlum.sh” failed to run.

```
root@84fc0790618d:~/demos/golang# ./run_golang_on_occlum.sh
/root/demos/golang/occlum_instance initialized as an Occlum instance
Enclave sign-tool: /opt/intel/sgxsdk/bin/x64/sgx_sign
Enclave sign-key: /opt/occlum/etc/template/Enclave.pem
SGX mode: HW
Building new image...
[+] Home dir is /root
[+] Open token file success!
[+] Token file invalid, will create new token file
[+] Init Enclave Successful 459561500674!
Building libOS...
EXPORT => OCCLUM_BUILTIN_CONF_FILE_MAC = 3a-b0-e4-30-29-68-f3-e0-06-e3-cb-8b-63-28-10-8e
EXPORT => OCCLUM_BUILTIN_VM_USER_SPACE_SIZE = 398458880
Signing the enclave...
<EnclaveConfiguration>
    <ProdID>0</ProdID>
    <ISVSVN>0</ISVSVN>
    <StackMaxSize>1048576</StackMaxSize>
    <HeapMaxSize>33554432</HeapMaxSize>
    <TCSNum>32</TCSNum>
    <TCSPolicy>1</TCSPolicy>
    <DisableDebug>0</DisableDebug>
    <MiscSelect>0</MiscSelect>
    <MiscMask>0xFFFFFFFF</MiscMask>
    <ReservedMemMaxSize>398458880</ReservedMemMaxSize>
    <ReservedMemMinSize>398458880</ReservedMemMinSize>
    <ReservedMemInitSize>398458880</ReservedMemInitSize>
    <ReservedMemExecutable>1</ReservedMemExecutable>
</EnclaveConfiguration>
tcs_num 32, tcs_max_num 32, tcs_min_pool 1
The required memory is 468226048B.
The required memory is 0x1be89000, 457252 KB.
Succeed.
Built the Occlum image and enclave successfully
occlum run /bin/web_server
/opt/occlum/build/bin/occlum: line 191:   179 Segmentation fault      (core dumped) RUST_BACKTRACE=1 "$instance_dir/build/bin/occlum-run" "$@"
root@84fc0790618d:~/demos/golang#

```

After enabling trace level logging, following messages come out:
```
root@84fc0790618d:~/demos/golang/occlum_instance# OCCLUM_LOG_LEVEL=trace occlum run /bin/web_server
[2020-08-20T05:07:31.287Z][TRACE][T0][#0] env_checked from env untrusted: []
[2020-08-20T05:07:31.287Z][TRACE][T0][#0] env_merged = ["OCCLUM=yes"]  (default env and untrusted env)
[2020-08-20T05:07:31.287Z][DEBUG][T0][#0] lookup_inode_follow: cwd: "/", path: "/bin/web_server"
[2020-08-20T05:07:31.585Z][DEBUG][T0][#0] lookup_inode_follow: cwd: "/", path: "/lib/ld-musl-x86_64.so.1"
[2020-08-20T05:07:31.597Z][DEBUG][T0][#0] allocated rsrv addr is 0x7fe466df5000, len is 0x17496000
[2020-08-20T05:07:31.640Z][ INFO][T0][#0] Process created: elf = /bin/web_server, pid = 1
[2020-08-20T05:07:31.640Z][TRACE][T1][#1][ArchPrctl] Syscall { num = ArchPrctl, code = 4098, addr = 0x7fe468e94d88 }
[2020-08-20T05:07:31.640Z][DEBUG][T1][#1][ArchPrctl] do_arch_prctl: code: ARCH_SET_FS, addr: 0x7fe468e94d88
[2020-08-20T05:07:31.640Z][TRACE][T1][#1][ArchPrctl] Retval = 0
[2020-08-20T05:07:31.640Z][TRACE][T1][#2][SetTidAddress] Syscall { num = SetTidAddress, tidptr = 0x7fe468e95364 }
[2020-08-20T05:07:31.640Z][DEBUG][T1][#2][SetTidAddress] set_tid_address: tidptr: 0x7fe468e95364
[2020-08-20T05:07:31.640Z][TRACE][T1][#2][SetTidAddress] Retval = 1
[2020-08-20T05:07:31.640Z][TRACE][T1][#3][Mprotect] Syscall { num = Mprotect, addr = 140618989375488, len = 4096, prot = 1 }
[2020-08-20T05:07:31.640Z][DEBUG][T1][#3][Mprotect] mprotect: addr: 0x7fe468e91000, size: 0x1000, perms: READ
[2020-08-20T05:07:31.640Z][ WARN][T1][#3][Mprotect] Do not support mprotect memory outside the mmap region yet
[2020-08-20T05:07:31.640Z][TRACE][T1][#3][Mprotect] Retval = 0
[2020-08-20T05:07:31.641Z][TRACE][T1][#4][Mprotect] Syscall { num = Mprotect, addr = 140618967400448, len = 6316032, prot = 1 }
[2020-08-20T05:07:31.641Z][DEBUG][T1][#4][Mprotect] mprotect: addr: 0x7fe46799c000, size: 0x606000, perms: READ
[2020-08-20T05:07:31.642Z][ WARN][T1][#4][Mprotect] Do not support mprotect memory outside the mmap region yet
[2020-08-20T05:07:31.642Z][TRACE][T1][#4][Mprotect] Retval = 0
[2020-08-20T05:07:31.642Z][TRACE][T1][#5][HandleException] Syscall { num = HandleException, info = 0x7fe464766ed0, fpregs = 0x7fe464766c80, context = 0x7e03c4bbdd3dfe77 }
[2020-08-20T05:07:31.642Z][DEBUG][T1][#5][HandleException] handle CPUID exception
[2020-08-20T05:07:31.642Z][TRACE][T1][#5][HandleException] cpuid result: CpuIdResult { eax: 22, ebx: 1970169159, ecx: 1818588270, edx: 1231384169 }
[2020-08-20T05:07:31.642Z][TRACE][T1][#5][HandleException] Retval = 0
[2020-08-20T05:07:31.642Z][TRACE][T1][#6][HandleException] Syscall { num = HandleException, info = 0x7fe464766ed0, fpregs = 0x7fe464766c80, context = 0x7e03c4bbdd3dfe77 }
[2020-08-20T05:07:31.642Z][DEBUG][T1][#6][HandleException] handle CPUID exception
[2020-08-20T05:07:31.642Z][TRACE][T1][#6][HandleException] cpuid result: CpuIdResult { eax: 591597, ebx: 118491136, ecx: 2147154943, edx: 3219913727 }
[2020-08-20T05:07:31.642Z][TRACE][T1][#6][HandleException] Retval = 0
[2020-08-20T05:07:31.642Z][TRACE][T1][#7][SchedGetaffinity] Syscall { num = SchedGetaffinity, pid = 0, cpusize = 8192, buf = 0x7fe46b493e58 }
[2020-08-20T05:07:31.642Z][DEBUG][T1][#7][SchedGetaffinity] do_sched_getaffinity tid: 0
[2020-08-20T05:07:31.642Z][TRACE][T1][#7][SchedGetaffinity] Retval = 2
/opt/occlum/build/bin/occlum: line 191:    45 Segmentation fault      (core dumped) RUST_BACKTRACE=1 "$instance_dir/build/bin/occlum-run" "$@"
root@84fc0790618d:~/demos/golang/occlum_instance#
root@84fc0790618d:~/demos/golang/occlum_instance#

```

Here is the container image info:

```
james@dr340s48:~$ sudo docker ps -a
CONTAINER ID        IMAGE                              COMMAND                  CREATED             STATUS                       PORTS               NAMES
16db419d7747        occlum/occlum:0.15.0-ubuntu18.04   "/bin/bash"              43 minutes ago      Exited (0) 36 minutes ago                        quirky_rubin

```

**Comment by dingelish** [comment]
more info:
```
$ uname -a
Linux dr340s48 4.15.0-109-generic #110-Ubuntu SMP Tue Jun 23 02:39:32 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
$ dpkg -l | grep sgx
ii  libsgx-ae-le                           2.10.100.2-bionic1                              amd64        Intel(R) Software Guard Extensions LE
ii  libsgx-aesm-launch-plugin              2.10.100.2-bionic1                              amd64        Launch Plugin for Intel(R) Software Guard Extensions AESM Service
ii  libsgx-enclave-common                  2.9.101.2-bionic1                               amd64        Intel(R) Software Guard Extensions Enclave Common Loader
ii  libsgx-launch                          2.10.100.2-bionic1                              amd64        Intel(R) Software Guard Extensions Launch Service
ii  libsgx-urts                            2.9.101.2-bionic1                               amd64        Intel(R) Software Guard Extensions uRTS
ii  sgx-aesm-service                       2.10.100.2-bionic1                              amd64        Intel(R) Software Guard Extensions AESM Service
```

**Comment by dingelish** [comment]
more info
driver: sgx_linux_x64_driver_2.6.0_95eaa6f.bin
```
$ uname -a
Linux dr340s48 4.15.0-109-generic #110-Ubuntu SMP Tue Jun 23 02:39:32 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
$ cat /etc/issue
Ubuntu 18.04.4 LTS \n \l
```

**Comment by jessehui** [comment]
We have reproduced this. This looks like we changed the syscall calling convention for libos but not change syscall calling convention for Go toolchains. We will fix that soon.

**Comment by guzongmin** [comment]
The issue fixed with https://github.com/occlum/occlum/commit/a151198d11eb9887531e3ea57d362c96fd3f53c9
Can you please try it?

**Comment by james-dong-security** [comment]
@guzongmin 
It is verified that latest source code with commit [a151198](https://github.com/occlum/occlum/commit/a151198d11eb9887531e3ea57d362c96fd3f53c9) does fix this reported issue.

Here are relevant logs:
===> server side
```
root@441ee3a38826:~/demos/golang#
root@441ee3a38826:~/demos/golang# ./run_golang_on_occlum.sh
/root/demos/golang/occlum_instance initialized as an Occlum instance
Enclave sign-tool: /opt/intel/sgxsdk/bin/x64/sgx_sign
Enclave sign-key: /opt/occlum/etc/template/Enclave.pem
SGX mode: HW
Building new image...
[+] Home dir is /root
[+] Open token file success!
[+] Token file invalid, will create new token file
[+] Init Enclave Successful 446676598786!
Building libOS...
EXPORT => OCCLUM_BUILTIN_CONF_FILE_MAC = 78-d7-28-ce-31-ca-43-1f-b6-25-dd-fd-86-48-a7-0f
EXPORT => OCCLUM_BUILTIN_VM_USER_SPACE_SIZE = 398458880
Signing the enclave...
<EnclaveConfiguration>
    <ProdID>0</ProdID>
    <ISVSVN>0</ISVSVN>
    <StackMaxSize>1048576</StackMaxSize>
    <HeapMaxSize>33554432</HeapMaxSize>
    <TCSNum>32</TCSNum>
    <TCSPolicy>1</TCSPolicy>
    <DisableDebug>0</DisableDebug>
    <MiscSelect>0</MiscSelect>
    <MiscMask>0xFFFFFFFF</MiscMask>
    <ReservedMemMaxSize>398458880</ReservedMemMaxSize>
    <ReservedMemMinSize>398458880</ReservedMemMinSize>
    <ReservedMemInitSize>398458880</ReservedMemInitSize>
    <ReservedMemExecutable>1</ReservedMemExecutable>
</EnclaveConfiguration>
tcs_num 32, tcs_max_num 32, tcs_min_pool 1
The required memory is 468226048B.
The required memory is 0x1be89000, 457252 KB.
Succeed.
Built the Occlum image and enclave successfully
occlum run /bin/web_server
[GIN-debug] [WARNING] Creating an Engine instance with the Logger and Recovery middleware already attached.

[GIN-debug] [WARNING] Running in "debug" mode. Switch to "release" mode in production.
 - using env:   export GIN_MODE=release
 - using code:  gin.SetMode(gin.ReleaseMode)

[GIN-debug] GET    /ping                     --> main.main.func1 (3 handlers)
[GIN-debug] Listening and serving HTTP on :8090
[GIN] 2020/08/26 - 05:30:12 | 200 |     162.525µs |       127.0.0.1 | GET      "/ping"
[GIN] 2020/08/26 - 05:30:19 | 200 |     392.305µs |       127.0.0.1 | GET      "/ping"

```

===> client side

```
root@441ee3a38826:~/demos/golang#
root@441ee3a38826:~/demos/golang# curl http://127.0.0.1:8090/ping
{"message":"pong"}root@441ee3a38826:~/demos/golang#
root@441ee3a38826:~/demos/golang#
root@441ee3a38826:~/demos/golang#
root@441ee3a38826:~/demos/golang#
root@441ee3a38826:~/demos/golang# curl http://127.0.0.1:8090/ping
{"message":"pong"}root@441ee3a38826:~/demos/golang#
root@441ee3a38826:~/demos/golang#
root@441ee3a38826:~/demos/golang#

```

**Comment by james-dong-security** [comment]
Fixed with commit [a151198](https://github.com/occlum/occlum/commit/a151198d11eb9887531e3ea57d362c96fd3f53c9).
